datafusion/optimizer/src/push_down_filter.rs

```diff
@@ -500,302 +387,359 @@ fn optimize_join(
-                    fetch: *fetch,
-                }),
-            )
-        }
-        _ => {
-            // all other plans are _not_ filter-commutable
-            let used_columns = plan
-                .schema()
-                .fields()
-                .iter()
-                .map(|f| f.qualified_column())
-                .collect::<HashSet<_>>();
-            issue_filters(state, used_columns, plan)
-        }
-    }
-}
+                push_down_all_join(predicates, filter.input(), left, right, vec![])?
+            }
+            LogicalPlan::TableScan(scan) => {
+                let mut new_scan_filters = scan.filters.clone();
+                let mut new_predicate = vec![];
+
+                let filter_predicates = utils::split_conjunction_owned(
+                    utils::cnf_rewrite(filter.predicate().clone()),
+                );
+
+                for filter_expr in &filter_predicates {
+                    let (preserve_filter_node, add_to_provider) =
+                        match scan.source.supports_filter_pushdown(filter_expr)? {
+                            TableProviderFilterPushDown::Unsupported => (true, false),
+                            TableProviderFilterPushDown::Inexact => (true, true),
+                            TableProviderFilterPushDown::Exact => (false, true),
+                        };
+                    if preserve_filter_node {
+                        new_predicate.push(filter_expr.clone());
+                    }
+                    if add_to_provider {
+                        // avoid reduplicated filter expr.
+                        if new_scan_filters.contains(filter_expr) {
+                            continue;
+                        }
+                        new_scan_filters.push(filter_expr.clone());
+                    }
+                }
 
-impl OptimizerRule for FilterPushDown {
-    fn name(&self) -> &str {
-        "filter_push_down"
-    }
+                let new_scan = LogicalPlan::TableScan(TableScan {
+                    source: scan.source.clone(),
+                    projection: scan.projection.clone(),
+                    projected_schema: scan.projected_schema.clone(),
+                    table_name: scan.table_name.clone(),
+                    filters: new_scan_filters,
+                    fetch: scan.fetch,
+                });
+
+                match conjunction(new_predicate) {
+                    Some(predicate) => LogicalPlan::Filter(Filter::try_new(
+                        predicate,
+                        Arc::new(new_scan),
+                    )?),
+                    None => new_scan,
+                }
+            }
+            _ => plan.clone(),
+        };
 
-    fn optimize(
-        &self,
-        plan: &LogicalPlan,
-        _: &mut OptimizerConfig,
-    ) -> Result<LogicalPlan> {
-        optimize(plan, State::default())
+        utils::optimize_children(self, &new_plan, optimizer_config)
```

Comment:
I see there was an `optimize_children()` call at the beginning of the method. 
Could you please explain why make a call to `optimize_children()` here again ?

````
let filter = match plan {
            LogicalPlan::Filter(filter) => filter,
            // we also need to pushdown filter in Join.
            LogicalPlan::Join(join) => {
                let optimized_plan = push_down_join(plan, join, None)?;
                return match optimized_plan {
                    Some(optimized_plan) => {
                        utils::optimize_children(self, &optimized_plan, optimizer_config)
                    }
                    None => utils::optimize_children(self, plan, optimizer_config),
                };
            }
            _ => return utils::optimize_children(self, plan, optimizer_config),
        };

````

Reply:
> I see there was an optimize_children() call at the beginning of the method.

Beginning place that called `optimize_children()` return directly. So they don't intersect.

Reply:
pattern is 

```
LogicalJoin -> push, optimize children, return directly

LogicalFilter
    SubPlan (Projection .....)

-> push, then optimize children

Other  ->  optimize children

```